### PR TITLE
feat(payments): move BVNK agreements to v1 sessions with checklist consent

### PR DESCRIPTION
**Pull request by multipletwigs**
**Summary**

- BVNK individual onboarding moves from the v2 agreement working set to v1 agreement sessions: mint on the residence step, sign on consent with the end user's IP, create the customer through `POST /platform/v1/customers` with the signed session reference, refresh status from the v1 customer GET everywhere (advance, requirements GET, customer webhooks).
- The consent requirement status is renamed `counterparty_collect_agreement` and carries the session's agreement name, display name, description, agreement URL and privacy-policy URL. Residence is US-only: BVNK publishes no v1 agreements for EU/EEA residents.
- The customer-link row is claimed at mint with SDP's `externalReference` alias, stamped `signedAt` on consent, and CAS-flipped to the v1 customer reference on create while keeping the residence country and session for the life of the row.
- `GET /v1/counterparties/{id}/provider-accounts` `customerLink` becomes a union on `provider`; the BVNK arm derives `providerStatus` (`PENDING_AGREEMENT` → `PENDING_DETAILS` → BVNK status), reports `providerCustomerReference: null` until the customer exists, and exposes `residenceCountryCode` and `agreements[]` with `signedAt`.
- Dashboard consent is a checklist gated in `useCounterpartyRequirements`: one row for the agreement, one for its privacy policy, both required before the primary action enables. The counterparty detail card shows the stage badge, residence flag, and per-agreement signed state with links.
- All v2 agreement code, schemas, fixtures, tests and the `bvnk:customers:agreements:status-change` webhook are removed. BVNK-supplied URLs are https-only at the zod boundary. Consent sign and the reference CAS outcome emit structured logs.

**before** — v2 agreement working set:

1. Residence → `POST /platform/v2/agreements` mints a working set keyed by our reference; presigned document URLs minted per requirements response.
2. Consent → `POST /platform/v2/agreements/actions` accepts each agreement; per-item ACCEPTED stored in `metadata.agreements.entries`.
3. Full pack → `POST /platform/v2/customers`; status refreshed from the v2 customer GET (`authenticatedLink` mandatory).
4. Customer link surfaced the `cp_…` alias as the provider customer reference with `providerStatus: null` before any customer existed.

**after** — v1 agreement sessions:

1. Residence → `POST /platform/v1/customers/agreement/sessions {customerType:"INDIVIDUAL", countryCode, useCase:"EMBEDDED_FIAT_ACCOUNTS"}`; static help-centre links stored in `metadata.session`.
2. Consent → `PUT …/sessions/{ref} {status:"SIGNED", ipAddress}` (204); `session.signedAt` stamped.
3. Full pack → `POST /platform/v1/customers {type:"individual", externalReference, signedAgreementSessionReference, individual}`; row reference CAS-flipped to the returned uuid, metadata keeps residence + session and gains `status`.
4. Status refresh → `GET /platform/v1/customers/{reference}`; Sumsub link only from `verification.url`. Customer link reports the true stage and a null reference until the customer exists.

## How the customer-link claim works

```text
mint     INSERT customer_link  provider_customer_reference = cp_<uuid>   metadata {residenceCountryCode, session{reference, agreements}}
consent  UPDATE metadata.session.signedAt
create   POST v1/customers {externalReference: cp_<uuid>, signedAgreementSessionReference} → {reference: R}
         UPDATE … SET provider_customer_reference = R, metadata = {…metadata, status}
                 WHERE provider_customer_reference = cp_<uuid>            (CAS; loser re-reads and converges on R)
```

## API before/after

`POST /v1/counterparties/{id}/requirements` — residence submitted (bvnk, onramp):

```jsonc
// before (reconstructed)
{ "status": "customer_agreement_required", "agreements": [{ "id": "…", "name": "…", "description": "…", "downloadUrl": "…" }] }
// after (live, local API)
{ "status": "counterparty_collect_agreement",
  "agreements": [{ "name": "EMBEDDED_PARTNER_PLATFORM_CUSTOMERS_US",
                   "displayName": "Embedded US Partner Platform Customers Agreement",
                   "description": "Embedded US Partner Platform Customers Agreement",
                   "url": "https://help.bvnk.com/hc/en-us/sections/…",
                   "privacyPolicyUrl": "https://help.bvnk.com/hc/en-us/articles/…" }] }
```

`POST /v1/counterparties/{id}/requirements` with `agreementConsent: true` → `{ "status": "collect_counterparty", "fields": [16 fields] }` (live).

`GET /v1/counterparties/{id}/provider-accounts` — `customerLink` after mint, then after consent (live):

```jsonc
// before (reconstructed)
{ "id": "…", "providerCustomerReference": "cp_bc2ce8cb40e8494bb47a7eb5f5d1ee55", "status": "active", "providerStatus": null, "createdAt": "…" }
// after mint
{ "provider": "bvnk", "id": "…", "providerCustomerReference": null, "status": "active",
  "providerStatus": "PENDING_AGREEMENT", "residenceCountryCode": "US",
  "agreements": [{ "name": "EMBEDDED_PARTNER_PLATFORM_CUSTOMERS_US", "displayName": "…", "url": "…", "privacyPolicyUrl": "…", "signedAt": null }] }
// after consent
{ …, "providerStatus": "PENDING_DETAILS", "agreements": [{ …, "signedAt": "2026-09-17T07:19:04.607Z" }] }
```

Unauthenticated `GET …/provider-accounts` → 401. Non-BVNK providers keep the flat shape with a non-null reference (`provider` added).

Webhook `POST /v1/webhooks/ramps/bvnk` — `bvnk:customers:agreements:status-change` is no longer handled (v2 working sets are gone); customer status events refresh through the v1 customer GET.

**Screenshots**

Counterparty information for BVNK agreements
CleanShot 2026-09-17 at 3 22 58 PM@2x

First, we show agreements based on tax residence. in V1 only US is supported.
CleanShot 2026-09-17 at 3 23 51 PM@2x

Secondly, we mint a session for the agreement claim for US based tax residences
CleanShot 2026-09-17 at 3 24 24 PM@2x

**Verification**

- `pnpm -C packages/sdp-types typecheck && pnpm -C packages/sdp-payments typecheck && pnpm -C apps/sdp-api typecheck && pnpm -C apps/sdp-web typecheck` — pass
- `pnpm exec biome check` on changed files — clean (one intentional `useOptionalChain` warning: `?.` is banned by house style)
- `pnpm -C apps/sdp-api playground:check` — pass; OpenAPI snapshot regenerated
- BVNK client `node:test` suite — 39/39
- local API: residence → `counterparty_collect_agreement`; consent → `collect_counterparty`; provider-accounts shows `PENDING_AGREEMENT` then `PENDING_DETAILS` with `signedAt` — as above
- local app: consent checklist and counterparty detail card exercised by the author
- Not run locally: DB-backed suites (`counterparties.test.ts`, `webhooks.test.ts`, repository CAS tests) and web unit tests — CI

Known gaps
- Consent `ipAddress` falls back to `0.0.0.0` when `getClientIp` resolves nothing (local, no trusted proxy). Temporary until client-IP plumbing lands.
- A crash between the v1 create and the row CAS relies on BVNK honoring `X-Idempotency-Key` replay on `POST /platform/v1/customers`; the sandbox probe only exercised a fresh key (400 `ACCOUNTS-2000`).
- No migration: existing BVNK test counterparties (v2-era rows without a stored session) fail the requirements and provider-accounts reads by design; create fresh counterparties.
- `bvnkCustomerVerificationRequirements` performs a second customer GET + row patch after the ensure step already refreshed it.

### CODE REVIEW

File: apps/sdp-api/src/routes/counterparty-provider-accounts/handlers.ts [hunk 1]
```diff
@@ -160,19 +163,88 @@ function mapProviderAccount(
...
+ * @returns Public BVNK customer-link object.
+ */
+function mapBvnkCustomerLink(
+  row: CounterpartyProviderAccountRow
+): BvnkCounterpartyProviderCustomerLink {
+  const metadata = bvnkCustomerProviderAccountMetadataSchema.parse(row.metadata);
+  const session = metadata.session;
+  const residenceCountryCode = metadata.residenceCountryCode;
+  if (session === undefined || residenceCountryCode === undefined) {
+    throw internalError("BVNK customer-link metadata is missing session state.");
```
[the remark this answers is not in the document]
Comment (multipletwigs) [reply apps/sdp-api/src/routes/counterparty-provider-accounts/handlers.ts]: Intentional. There are no users on devnet or mainnet yet, so the v2-era customer-link rows are test data and are abandoned rather than migrated: the PR ships no data migration and fresh counterparties are created instead (listed under Known gaps). A BVNK row without a stored session is not a state the v1 flow can produce, so the mapper treats it as an invariant violation instead of carrying a compatibility branch.

[the remark this answers is not in the document]
Comment (multipletwigs) [reply apps/sdp-api/src/routes/counterparty-provider-accounts/handlers.ts]: Resolving as intentional, and cf3cb72c5 tightens it rather than adding compat. The only pre-customer row shape the v1 flow can leave without a `session` is a claimed-but-unminted row (crash or concurrent window between the claim upsert and the mint); that now maps to `PENDING_AGREEMENT` with `agreements: []`. A row with a v1 `status` and no `session` cannot be produced by this code, so it stays an invariant violation. v2-era rows are devnet test data and are recreated, not migrated.

File: packages/sdp-types/src/ramp-destinations.ts [hunk 2]
```diff
@@ -26,6 +26,13 @@ export const MOONPAY_HOSTED_APPROVED_HOSTS = [
 /** Coinbase headless-onramp payment-link host (sandbox uses the same host). */
 export const COINBASE_HOSTED_APPROVED_HOSTS = ["pay.coinbase.com"] as const;
 
+/**
+ * BVNK customer-link hosts: the agreement and privacy-policy pages. The Sumsub
+ * verification link is opened via `openExternalRampUrl` instead and its host is
+ * not pinned by design.
+ */
+export const BVNK_CUSTOMER_LINK_APPROVED_HOSTS = ["help.bvnk.com"] as const;
```
[the remark this answers is not in the document]
Comment (multipletwigs) [reply packages/sdp-types/src/ramp-destinations.ts]: Fixed in cf3cb72c5. `BVNK_CUSTOMER_LINK_APPROVED_HOSTS` is now `["help.bvnk.com", "www.bvnk.com"]`, and `bvnk-agreement-consent.unit.test.tsx` clicks both the agreement and privacy-policy links and asserts `window.open` is called with each URL, `_blank`, `noopener`.

File: apps/sdp-api/src/routes/payments/handlers/ramps/bvnk.ts [hunk 3]
```diff
@@ -839,74 +841,76 @@ export async function ensureBvnkCustomer(
...
-    agreements.id,
-    entries,
-    residenceCountry
-  );
-  const pending = agreements.agreements.filter((agreement) => agreement.status !== "ACCEPTED");
-  if (pending.length > 0) {
-    return { requirements: await bvnkAgreementDetails(c, direction, agreements) };
-  }
-  return { requirements: bvnkCollectCounterparty(direction, residenceCountry) };
+  const session = await client.createAgreementSession(ctx, { countryCode: residenceCountry });
```
[the remark this answers is not in the document]
Comment (multipletwigs) [reply apps/sdp-api/src/routes/payments/handlers/ramps/bvnk.ts]: Fixed in cf3cb72c5 with claim-before-spend. Residence submit first claims the customer-link row (the existing upsert, jsonb-merge so it never clobbers a stored session), then mints with `X-Idempotency-Key` = the row uuid, then CAS-writes the session via `setCustomerLinkSession` (`WHERE NOT (metadata ? 'session')`). A concurrent mint loses the CAS, re-reads, and returns the winner's session. A crash between claim and mint leaves a `session_pending` row that the next submit mints from with the same key. Covered by repository CAS hit/miss/tenant tests and route tests for claimed-unminted, already-stored (zero BVNK calls), and the header value.

File: apps/sdp-api/src/routes/payments/handlers/ramps/bvnk.ts [hunk 4]
```diff
@@ -516,166 +453,322 @@ function requireBvnkResidenceCountry(metadata: BvnkCustomerProviderAccountMetada
...
-    organizationId: counterparty.organization_id,
-    projectId,
+    ...scope,
     customer: {
-      customerReference: workingSetId,
-      residenceCountryCode,
-      agreements: { entries },
+      customerReference: buildBvnkCustomerExternalReference(input.counterparty.id),
+      residenceCountryCode: input.residenceCountry,
+    },
```
[the remark this answers is not in the document]
Comment (multipletwigs) [reply apps/sdp-api/src/routes/payments/handlers/ramps/bvnk.ts:636]: Fixed in 94b65da35. The claim upsert now merges as `EXCLUDED.metadata || stored`, so stored keys win and a concurrent claim cannot replace the residence country. The mint then reads the stored country, rejects a submit naming a different one with 409 before any BVNK call, and always mints the session for the stored country, so residence and session on a row always come from the same claim. Repository test covers the repeat claim keeping the first country; route test covers the divergent submit (409, zero provider calls, row untouched).

**Review by resourcefulmind — APPROVED** [review]
